### PR TITLE
[BEAM-948] Add ability to write files to GridFS

### PR DESCRIPTION
**Pull request by dkulp**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by jbonofre** [comment]
R: @jbonofre

File: sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbGridFSIO.java [hunk 1]
```diff
@@ -446,4 +504,164 @@ public void close() throws IOException {
...
+  }
+
+  /**
+   * A {@link PTransform} to write data to MongoDB GridFS.
+   */
+  @AutoValue
+  public abstract static class Write<T> extends PTransform<PCollection<T>, PDone> {
+    @Nullable abstract String uri();
+    @Nullable abstract String database();
+    @Nullable abstract String bucket();
```
Comment (jbonofre) [inline sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbGridFSIO.java]: As these attributes are also used in the `Read`, I would recommend to create a `ConnectionConfiguration` POJO (using `AutoValue`) used by the `Read` and `Write`.

File: sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbGridFSIO.java [hunk 2]
```diff
@@ -446,4 +504,164 @@ public void close() throws IOException {
...
+
+    public GridFsWriteFn(Write<T> spec) {
+      this.spec = spec;
+    }
+
+    private Mongo setupMongo() {
+      return spec.uri() == null ? new Mongo() : new Mongo(new MongoURI(spec.uri()));
+    }
+
+    private GridFS setupGridFS(Mongo mongo) {
```
Comment (jbonofre) [inline sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbGridFSIO.java]: This method could be in the `Connection` POJO used by both `Read` and `Write` according to the previous comment.

File: sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbGridFSIO.java [hunk 3]
```diff
@@ -446,4 +504,164 @@ public void close() throws IOException {
...
+
+    @ProcessElement
+    public void processElement(ProcessContext context) throws Exception {
+      T record = context.element();
+      spec.writeFn().write(record, outputStream);
+    }
+
+    @FinishBundle
+    public void finishBundle(Context context) throws Exception {
+      if (gridFsFile != null) {
```
Comment (jbonofre) [inline sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbGridFSIO.java]: The bundle can "live" for a long time. Are you sure you want to flush/close the stream at the end of the bundle ? The stream can be open for a long time.

Comment (dkulp) [inline sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbGridFSIO.java]: That should be fine.  Whenever the "chunkSize" is hit (which is configurable), the chunk is persisted to the database and file updated and such.

**Comment by dkulp** [comment]
Jenkins failure seems completely unrelated.

**Comment by jbonofre** [comment]
Yes, agree. I'm building the branch locally anyway.

**Comment by jbonofre** [comment]
By the way, not directly related, but it would be great to avoid deprecated code when possible:
* `MongoClient` instead of `Mongo` (deprecated)
* `MongoClientURI` instead of `MongoURI`
* `getDatabase()` instead of `getDB()`

It's used in the `MongoDbIO`, so `MongoDbGridFSIO` could use it as well.

**Comment by jbonofre** [comment]
Can you please rebase (I did a change on the test) ? Else I can do it for you.

**Comment by dkulp** [comment]
Working on the rebase... changes to the direct runner apparently are changing how the bundles are created so the new test is failing.  I'll work on that tomorrow.

Regarding the deprecated classes:  there is nothing that can done right now.   The GridFS API's require the use of the older classes.   The new 3.3 version of the mongo drivers has new GridFS API's that can use the new classes, but I haven't seen anyone yet willing to use the 3.3 drivers in production.   To support 3.2 and older, we have to use the deprecated objects.

**Comment by jbonofre** [comment]
Thanks, no problem for deprecated classes.

I will take a look on the failing test as well.

**Comment by dkulp** [comment]
Rebase complete, should be good to go.

**Comment by jbonofre** [comment]
Thanks, waiting the end of Jenkins execution.

**Comment by jbonofre** [comment]
LGTM

**Comment by jbonofre** [comment]
I gonna merge.

**Pull request closed without merging**